xdem/coreg.py

```diff
@@ -1205,16 +1227,79 @@ def apply_pts(self, coords: np.ndarray) -> np.ndarray:
 
         :returns: The transformed coordinates.
         """
-        if not self._fit_called:
+        if not self._fit_called and self._meta.get("matrix") is None:
             raise AssertionError(".fit() does not seem to have been called yet")
         assert coords.shape[1] == 3, f"'coords' shape must be (N, 3). Given shape: {coords.shape}"
 
-        return self._apply_pts_func(coords)
+        # See if an _apply_pts_func exists
+        try:
+            transformed_points = self._apply_pts_func(coords)
+        # If it doesn't exist, use opencv's perspectiveTransform
+        except NotImplementedError:
+            if self.is_rigid:  # This only works on it's rigid, however.
+                transformed_points = cv2.perspectiveTransform(coords.reshape(1, -1, 3), self.to_matrix()).squeeze()
+            else:
+                raise ValueError("Coreg method is non-rigid but has not implemented _apply_pts_func")
+
+        return transformed_points
+
+    @property
+    def is_rigid(self) -> bool:
+        """Check if the transform be explained by a rigid transform."""
+        # _is_rigid is found by seeing if to_matrix() raises an error.
+        # If this hasn't been done yet, it will be None
+        if self._is_rigid is None:
+            try:  # See if to_matrix() raises an error.
+                self.to_matrix()
```

Comment:
I understand from this that your definition of 'rigid' is that it an be expressed as a 4x4 matrix?
Scaling can be expressed using a 4x4 matrix, but is not a rigid transformation, so I'm not sure the term is appropriate here. Or did I miss something?

Reply:
Good point, `is_rigid` doesn't work... What about `is_helmert` or something (after the [Helmert transform](https://en.wikipedia.org/wiki/Helmert_transformation)). Open to other suggestions!

What it should mean is exactly what you ask; "can it be described as a 4x4 matrix?"

Reply:
Or `is_affine`?

Reply:
Source: https://www.euclideanspace.com/maths/geometry/affine/matrix4x4/index.htm

Reply:
Hmm, maybe that could work... It just shouldn't be mixed up with 2D affine transforms, like `rio.transform.Affine`, which is very different from a 3D "affine" transform.

Reply:
Changed `is_rigid` to `is_affine` in be9e9fa.